File: Client/Gamify.Client.Net/Gamify.Client.Net/Managers/GameCreationManager.cs
using Gamify.Client.Net.Client;
using Gamify.Client.Net.Services;
using Gamify.Contracts.Notifications;
using Gamify.Contracts.Requests;
using System;

namespace Gamify.Client.Net.Managers
{
    public class GameCreationManager
    {
        private readonly IGameService<CreateGameRequestObject, GameInviteNotificationObject> createGameService;
        private readonly IGameService<GameAcceptedRequestObject, GameCreatedNotificationObject> acceptGameService;
        private readonly IGameService<GameRejectedRequestObject, GameRejectedNotificationObject> rejectGameService;

        public event EventHandler<GameNotificationEventArgs<GameInviteNotificationObject>> GameInviteNotificationReceived;
        public event EventHandler<GameNotificationEventArgs<GameCreatedNotificationObject>> GameCreatedNotificationReceived;
        public event EventHandler<GameNotificationEventArgs<GameRejectedNotificationObject>> GameRejectedNotificationReceived;

        public GameCreationManager(string playerName, IGameClientFactory clientFactory)
        {
            var gameClient = clientFactory.GetGameClient(playerName);

            this.createGameService = new GameService<CreateGameRequestObject, GameInviteNotificationObject>(GameRequestType.CreateGame, GameNotificationType.GameInvite, gameClient);
            this.acceptGameService = new GameService<GameAcceptedRequestObject, GameCreatedNotificationObject>(GameRequestType.GameAccepted, GameNotificationType.GameCreated, gameClient);
            this.rejectGameService = new GameService<GameRejectedRequestObject, GameRejectedNotificationObject>(GameRequestType.GameRejected, GameNotificationType.GameRejected, gameClient);

            this.createGameService.NotificationReceived += (sender, args) =>
            {
                this.NotifyGameInvite(args);
            };
            this.acceptGameService.NotificationReceived += (sender, args) =>
            {
                this.NotifyGameCreated(args);
            };
            this.rejectGameService.NotificationReceived += (sender, args) =>
            {
                this.NotifyGameRejected(args);
            };
        }

        public void CreateGame(CreateGameRequestObject createGameRequest)
        {
            this.createGameService.Send(createGameRequest);
        }

        public void AcceptGame(GameAcceptedRequestObject acceptGameGameRequest)
        {
            this.acceptGameService.Send(acceptGameGameRequest);
        }

        public void RejectGame(GameRejectedRequestObject rejectGameGameRequest)
        {
            this.rejectGameService.Send(rejectGameGameRequest);
        }

        private void NotifyGameInvite(GameNotificationEventArgs<GameInviteNotificationObject> args)
        {
            if (this.GameInviteNotificationReceived != null)
            {
                this.GameInviteNotificationReceived(this, args);
            }
        }

        private void NotifyGameCreated(GameNotificationEventArgs<GameCreatedNotificationObject> args)
        {
            if (this.GameCreatedNotificationReceived != null)
            {
                this.GameCreatedNotificationReceived(this, args);
            }
        }

        private void NotifyGameRejected(GameNotificationEventArgs<GameRejectedNotificationObject> args)
        {
            if (this.GameRejectedNotificationReceived != null)
            {
                this.GameRejectedNotificationReceived(this, args);
            }
        }
    }
}

File: Client/Java/gamify-java-client/src/main/java/ar/com/mdf/gamify/managers/GameCreationManager.java
package ar.com.mdf.gamify.managers;


import java.net.URISyntaxException;

import ar.com.mdf.gamify.client.IGamifyClient;
import ar.com.mdf.gamify.enums.NotificationType;
import ar.com.mdf.gamify.enums.RequestType;
import ar.com.mdf.gamify.exceptions.NotConnectedException;
import ar.com.mdf.gamify.factory.IClientFactory;
import ar.com.mdf.gamify.objects.notification.InvitationToGameNotificationObject;
import ar.com.mdf.gamify.objects.notification.ResponseInvitationGameNotificationObject;
import ar.com.mdf.gamify.objects.request.AcceptGameRequestObject;
import ar.com.mdf.gamify.objects.request.SessionRequestObject;
import ar.com.mdf.gamify.objects.request.StartGameRequestObject;
import ar.com.mdf.gamify.observers.IGameObserver;
import ar.com.mdf.gamify.services.GameService;
import ar.com.mdf.gamify.services.IGameService;

public class GameCreationManager {
	
	private IGameService<StartGameRequestObject, InvitationToGameNotificationObject> createGameService;
	private IGameService<AcceptGameRequestObject, ResponseInvitationGameNotificationObject> acceptGameService;
	private IGameService<SessionRequestObject, ResponseInvitationGameNotificationObject> rejectGameService;
	private IGamifyClient client;
	
	public GameCreationManager(IClientFactory clientFactory) throws NotConnectedException {
		try {
			this.client = clientFactory.getClient();
		} catch (URISyntaxException e) {
			throw new NotConnectedException(e);
		}
		this.createGameService = new GameService<StartGameRequestObject, InvitationToGameNotificationObject>(RequestType.CreateGame, NotificationType.GameInvite, this.client, InvitationToGameNotificationObject.class);
		this.acceptGameService = new GameService<AcceptGameRequestObject, ResponseInvitationGameNotificationObject>(RequestType.GameAccepted, NotificationType.GameCreated, this.client,ResponseInvitationGameNotificationObject.class);
		this.rejectGameService = new GameService<SessionRequestObject, ResponseInvitationGameNotificationObject>(RequestType.GameRejected, NotificationType.GameRejected, this.client, ResponseInvitationGameNotificationObject.class);
	}
	
	public void suscribeCreateGame(IGameObserver<InvitationToGameNotificationObject> observer){
		this.createGameService.suscribe(observer);
	}
	public void suscribeAcceptGame(IGameObserver<ResponseInvitationGameNotificationObject> observer){
		this.acceptGameService.suscribe(observer);
	}
	public void suscribeRejectGame(IGameObserver<ResponseInvitationGameNotificationObject> observer){
		this.rejectGameService.suscribe(observer);
	}

	public void createGame(StartGameRequestObject inviteRequest) {
		createGameService.send(inviteRequest);
	}
	public void acceptGame(AcceptGameRequestObject acceptGameRequest) {
		acceptGameService.send(acceptGameRequest);
	}
	public void rejectGame(SessionRequestObject rejectGameRequest) {
		rejectGameService.send(rejectGameRequest);
	}

}
